File: 234. Palindrome Linked List.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func isPalindrome(head *ListNode) bool {
    mid := getMiddle(head)
    reversed := reverse(mid.Next)
    for head != nil && reversed != nil {
        if (head.Val != reversed.Val) {
            return false
        }
        head = head.Next
        reversed = reversed.Next
    }
    return true
}

func reverse(head *ListNode) *ListNode {
    var previous *ListNode
    for head != nil {
        temp := head.Next
        head.Next = previous
        previous= head
        head = temp
    }
    return previous
}

func getMiddle(head *ListNode) *ListNode {
    length := getLength(head)
    for length > 0 {
        head = head.Next
        length -= 1
    }
    return head
}

func getLength(head *ListNode) int {
    length := 0
    for head != nil {
        length += 1
        head = head.Next
    }
    return length/2-1
}

File: 234. Palindrome Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        ListNode mid = getMiddle(head);
        ListNode reversed = reverse(mid.next);
        while (head != null && reversed != null) {
            if (head.val != reversed.val) return false;
            head = head.next;
            reversed = reversed.next;
        }
        return true;
    }
    private ListNode reverse(ListNode head) {
        ListNode previous = null;
        while (head != null) {
            ListNode temp = head.next;
            head.next = previous;
            previous = head;
            head = temp;
        }
        return previous;
    }
    private ListNode getMiddle(ListNode head) {
        int length = getLength(head);
        while (length-- > 0) head = head.next;
        return head;
    }
    private int getLength(ListNode head) {
        int length = 0;
        while (head != null) {
            length++;
            head = head.next;
        }
        return length/2-1;
    }
}

